### ISSUE TITLE
A speech board / speech mode for blind users

### ISSUE THREAD

**Issue by cjb** [OptiKey/OptiKey#242]
Hi!  I'm a programmer who has volunteered to help out a speech therapist who has an unusual use case for OptiKey -- helping someone to communicate who is lacking motor control and blind, but has control over eye movement.

It sounds like the mode would involve having large buttons covering each compass edge of the screen, each one representing a word like Yes/No/Help.  Instead of adding the chosen word to a text box, it would just be spoken aloud straight away.  It doesn't need to be recorded anywhere.

Please could you give me some technical advice on a good (maybe even mergeable!) way to implement this, starting with some build instructions for OptiKey?  I'm afraid I don't usually work on Windows/.NET, so it'll take me a little while to get up to speed.

It looks like `Menu.xaml` has some visually similar layout to this idea, so I'm guessing I could work on a layout by modifying that.  Once the layout is done, the other task is having speech happen after each symbol is chosen rather than waiting for the Speak symbol to be chosen.  Any hints would be great, thanks!

**Comment by JuliusSweetland** [comment]
@cjb Hi Chris. Apologies for the delay in getting back to you - lots going on at the moment.

You can achieve a proof of concept with only a few changes:
* Create a new keyboard with 2 keys on it to get started, e.g. "Yes" and "No". As the user is partially/completely blind these need to be as large as possible to cover the left and right halves of the screen, for example.
* Make this keyboard the startup keyboard and make sure OptiKey starts fullscreen, which currently happens when you are in "Conversation Mode".
* Change MainViewModel to make direct calls to the AudioService to generate speech, instead of passing the selection to the KeyboardOutputService for handling.
* Run it with an eye tracker as the point and trigger source and try making yes/no selections. How does it feel?

If everything is working then you could strip back all the unnecessary stuff in OptiKey, which off the top of my head would include:
* Simplify the InputService to remove the MultiKeySelection logic as this won't be needed.
* Simplify the MainViewModel (note the file is in 2 parts or "partials") - you can remove all handling of mouse events and all FunctionKeys values that you won't use. Also you could remove all handling of 'Text' as this is the scratch pad, which you don't need.
* You could simplify the Views to remove most of the keyboards - I presume you'll be creating new, simplified keyboards. You won't need the scratch pad either.
* You could probably remove the whole KeyboardOutputService.
* You could remove all localisation stuff, including the resource files and logic in the KeyboardHost.

**Comment by JuliusSweetland** [comment]
@cjb Another issue has just been created (https://github.com/OptiKey/OptiKey/issues/244) which is asking for a very similar thing; a sort of speech board with no fancy features. As this overlaps with what you are doing I'm going to close the other issue and point the creator here.

I had an idea of how this change could work for both of you; once you have a proof of concept it could be extended to read a text file with one key per line. The entry is either text, e.g. "I am hungry", or an existing function key, e.g. Sleep. OptiKey reads the file and generates a View which lays out the keys as large as possible and then runs as normal. This would allow users to customise their own keyboard. What do you think?

Also, how are you getting on?

**Comment by PiotrKurnik** [comment]
Hi guys, 

Indeed issue that I have raised is the same as the one described by @cjb. I have zero experience with Windows/.NET and therefore I still figure out the building blocks of that. 
@JuliusSweetland your concept seems to be fixing all the problems, at least I think it's very promising. I will do my best to follow your description and produce a simple keyboard. I just have to figure out what do I need to achieve this task.

**Comment by cjb** [comment]
Hi!  Just an update: I'm stalled waiting for some feedback on whether a simple test of the idea worked well with the client it's being developed for.  So please don't wait on me for anything.  :)

**Comment by JuliusSweetland** [comment]
No rush. I've not got any spare time either at the moment!

**Comment by alexandre-mbm** [comment]
@cjb, initial examples from eye trackers's SDKs could work. Why not? Changing a large project such as OptiKey appears me much more complex.

**Comment by vijayrajanna** [comment]
@cjb, I just saw this thread and it's an interesting use case.
As @JuliusSweetland has already mentioned, making those changes will help you achieve the desired functionality. Also, understand that you may not have much experience with C# .net.

While this feature can be implemented, I have a question for you.
How can you get the system calibrated with blind users, it's a hard problem.
Do you have a suggestion?

Recently, I'm developing a system to aid in calibration with blind users.
I you are still interested, I am open for discussions.
Thanks.

**Comment by JuliusSweetland** [comment]
Closing due to lack of progress. This can be re-opened if required.
